zerver/tests/test_upload.py

```diff
@@ -129,9 +129,39 @@ def test_file_upload_authed(self):
         self.send_message("hamlet@zulip.com", "Denmark", Recipient.STREAM, body, "test")
         self.assertIn('title="zulip.txt"', self.get_last_message().rendered_content)
 
-    def test_delete_old_unclaimed_attachments(self):
+    def test_file_download_unauthed(self):
         # type: () -> None
+        self.login("hamlet@zulip.com")
+        fp = StringIO("zulip!")
+        fp.name = "zulip.txt"
+        result = self.client.post("/json/upload_file", {'file': fp})
+        json = ujson.loads(result.content)
+        uri = json["uri"]
+
+        self.client.post('/accounts/logout/')
+        response = self.client.get(uri)
+        self.assert_json_error(response, "Not logged in: API authentication or user session required",
+                                status_code=401)
+
+    def test_removed_file_download(self):
+        # type: () -> None
+        '''
+        Trying to download deleted files should return 404 error
+        '''
+        self.login("hamlet@zulip.com")
+        fp = StringIO("zulip!")
+        fp.name = "zulip.txt"
+        result = self.client.post("/json/upload_file", {'file': fp})
+        json = ujson.loads(result.content)
+        uri = json["uri"]
 
+        destroy_uploads()
+
+        response = self.client.get(uri)
+        self.assertEqual(response.status_code, 404)
```

Comment:
@timabbott So 2 things can happen:

- The user requests for a file that was never uploaded. Here we return a json_error
- The user requests for a file that was deleted but an entry still exists in the database. Here an HttpResponseNotFound is returned.

Ive added a seperate test for the first case.